### PR TITLE
Proposal for improvements to concrete notation

### LINKED ISSUE #422: Use `abs x. tactic` instead of `<x> tac`.
I think it is more uniform to have
```
lam x.
let x = ....
abs x.
```
to introduce bindings. Thoughts?

### PR DESCRIPTION
**Pull request by jonsterling**
This doesn't get close to addressing the confusing-ness, but I think that this contains some improvements that will make things more consistent and less confusing.

Resolves #422

### CODE REVIEW

**Comment by favonia** [comment]
Random idea: would `claim` match mathematical writings better?

**Comment by jonsterling** [comment]
I kind of like that idea!

**Comment by favonia** [comment]
Do you have ideas about the syntax in #404 in the new paradigm? And sorry for destroying the mergability.

**Comment by jonsterling** [comment]
@favonia I don't have any thoughts yet about that, but I'm open to ideas. I think #404 is a good idea.

**Comment by jonsterling** [comment]
And no worries about the merge conflicts 😄 

@cangiuli @favonia @mortberg Do you think that this PR is generally an improvement? If so, I can merge it and then update any other branches that are currently open. Also happy to hold off on merging this until another time.

**Comment by favonia** [comment]
I think the most difficult part is to remember when to use `in` and when to use `=>`. The rule I figured out is that if it results into a binder in the realizer, then it's `=>`. Otherwise, it's `in`. Is that right? Following this rule, I might suggest
```
fresh x -> blah in blah
```
instead of
```
fresh x in blah; blah
```

**Comment by jonsterling** [comment]
@favonia Yes, you're right that this is still confusing... At first I thought to use `=>` for everything! That would be quite consistent, but I was worried it might be a bit surprising.

I think the rule you figured out may hold, but it is not intentional (at least, I wasn't thinking of it when I came up with these notations). I was more thinking about how in languages (like OCaml) we use an arrow for function binders, but `in` for let bindings, and I tried to extrapolate from there.

**Comment by favonia** [comment]
Let me list the notations for a clearer view, as a precursor of #455...
```
lam x y z =>
abs x y z =>
```
How about semicolons for all of these, then?
```
claim lemma : [ .... ] by [ .... ]; ...
let x = ...; ... (very Haskell-ish)
query x <- loc; ...
fresh x -> tac; ... (???)
```
Also I think
```
a, b | #jdg{%a = %b in bool}
```
should be replaced by
```
[[a b] #jdg{%a = %b in bool} => tac]
```
or something.

**Comment by jonsterling** [comment]
@favonia The reason I don't want to use the semicolon for these is that the semicolon already has a meaning on its own: it is a tactical. Since these tacticals cannot really be construed as instances of the semicolon, it would lead to confusion I think.

**Comment by jonsterling** [comment]
I like your idea about improving the match syntax!

**Comment by jonsterling** [comment]
But maybe the semicolon is OK, perhaps users don't need to understand the difference. What do you think?

**Comment by favonia** [comment]
Are you talking about `;` being the sequencing operation?

**Comment by jonsterling** [comment]
Yeah

**Comment by favonia** [comment]
I don't know. These notational issues never bother me too much. I haven't thought about the semantics carefully... perhaps you could elaborate more on the important differences in your mind? In the name of god of S-expressions, I think they can be "disguised" as sequencing if we really want to. Suppose we finally decide to use `(seq [x y z] tac1 [w] tac2)` as the ultimate proof construct:
```
[lemma] (claim [ .... ] [ .... ])
[x] (...?...) // let
[x] (query loc)
[x] (elim y)
```
Note that `(seq [x] tac1 [y] tac2)` takes the freedom to break the usual binding structure. It should be parsed into some ABT that makes sense. See #395.

**Comment by jonsterling** [comment]
@favonia One important difference is that in `t1; t2`, `t2` is applied to all the subgoals generated by `t1`. But in `lam x; t`, `t` would be applied only to one of two subgoals generated by `lam x`. So, it disrupts the semantics of the semicolon and makes it a little harder to understand what's going on.

Maybe it doesn't really matter... I have a hard time figuring out which kinds of deviations are most likely to cause confusion among users.

**Comment by jonsterling** [comment]
With that said, I'm leaning toward finding a way to escape from the `seq` (which is currently written using that `fresh` thing) being a binding operator... It is kind of nice for modularity of proofs, but it has also caused a lot of complication. I don't yet have a proposal for resolving that at the moment, so feel free to ignore this remark.

**Comment by favonia** [comment]
I think the difference you mentioned is clear and I did not suggest semicolons for `lam` or `abs`. How about other tacticals?

**Comment by jonsterling** [comment]
> I think the difference you mentioned is clear and I did not suggest semicolons for lam or abs. How about other tacticals?

Oh, then I misread your comment! Sorry. I think it makes sense the ones you actually suggested. 😄  Your proposal has the benefit of making it so that `in` is only ever used for one thing, which is to select targets like in `reduce` or `unfold`!

**Comment by favonia** [comment]
> `[[a b] #jdg{%a = %b in bool} => tac]`

Actually now I'm not sure about `[a b]` which is very S-expression-with-RedPRL-flavor but we are not faithful S-expressionists yet.

**Comment by jonsterling** [comment]
> Actually now I'm not sure about [a b] which is very S-expression-with-RedPRL-flavor. We are not faithful S-expressionists yet.


Maybe a compromise for now would be to at least get rid of the commas, since I have got rid of commas from the other things.

File: test/success/equality-elim.prl [hunk 1]
```diff
@@ -1,16 +1,16 @@
 Tac GetHole(#c : [exp].exp, #t : [exp].tac) = [
-  query gl <- concl.
+  query gl <- concl;
   match gl {
     [hole | #jdg{(#c %hole)} => (#t %hole)]
   }
 ].
 
-// We can write a cool user-defined tactic for asserting and then rewriting along an equality.
+// We can write a cool user-defined tactic for claiming and then rewriting along an equality.
```
Comment (jonsterling) [inline test/success/equality-elim.prl]: lol autocorrect 😎

**Comment by jonsterling** [comment]
@favonia I think I've integrated your Thought now.

**Comment by favonia** [comment]
Since we are changing the grammar, perhaps we can kill #376 as well.

**Comment by jonsterling** [comment]
@favonia good idea! done.

File: test/success/hlevels.prl [hunk 2]
```diff
@@ -38,19 +38,15 @@ Thm PropSet(#l:lvl) : [
...
-  (abs i => lam a b =>
-    let foo = (PropSet #l) [`tyA, `h1];
-    unfold IsSet IsProp;
-    use foo [`a, `b, use h1 [`a, `b], use h2 [`a, `b], `i]);
-  unfold PropSet; auto
+  abs i => lam a b =>
+    `(@ ($ (PropSet #l) tyA h1 a b ($ h1 a b) ($ h2 a b)) i)
 ].
 
 Thm IsPropIsProp'(#l:lvl) : [
```
Comment (favonia) [inline test/success/hlevels.prl]: Wait, did you see `IsPropIsProp'` below?

Comment (favonia) [reply test/success/hlevels.prl]: Perhaps we can delete `IsPropIsProp'` now.

Comment (jonsterling) [reply test/success/hlevels.prl]: oops, I mindlessly didn't see that

Comment (jonsterling) [reply test/success/hlevels.prl]: Well, I'll leave both in because the first one indicates that there are some kinks to be worked out in the tactics.

**Comment by jonsterling** [comment]
I want to fix some parser issues (like #469 and #468) from *this* branch rather than from `master`, to avoid duplicating work.

@cangiuli To avoid causing trouble for you, there are a couple alternatives:

1. We could merge your partial `UA` work into master, and then I can rebase this onto that and update your code.

2. We could merge this first, but then I can quickly bring your branch up to date.

3. If you don't like this branch, we could abandon it or try again later.

**Comment by cangiuli** [comment]
@jonsterling Feel free to merge my work into master. If you comment out `UARetSig` then the file should work. You can also squash my commits if you want; they're not useful and my rebase also caused some duplication.

**Comment by jonsterling** [comment]
@cangiuli OK! Thanks. I'll take care of it!

**Pull request merged**